### ISSUE TITLE
Longevity 1000 keyspaces, a 90% drop of throughput while replacing a node

### ISSUE THREAD

**Issue by fruch** [scylladb/scylla#4981]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 3.1.0.rc5-0.20190902.623ea5e3d
Cluster size: 6 
OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-027fe33bfba73c292

# Summery

![image](https://user-images.githubusercontent.com/340979/64514876-371cd380-d2f4-11e9-8c10-d3159f9ee645.png)

While nemesis terminate_and_replace was running, a %90 drop in throughput was noticed, and all nodes in cluster got to 100% cpu (with the exception of the node being added) 

# Logs
* **monitor snapshot** - https://snapshot.raintank.io/dashboard/snapshot/zzyLc4k7y5RnpKsEH1xEUWFUwHWzUZYv

* **all nodes log** http://scratch.scylladb.com/fruch/1000_ks_terminate_and_replace_throughput_drop.tar.gz

NOTE: in those log adding of the new node isn't yet finalized (i.e. the nemesis didn't yet finished)

**Comment by nyh** [comment]
This is **not** the same issue that happens on a simple node reboot, which comes up with an empty cache and slows everything down. In such a case you would see low performance, but also low load on most nodes. What we see here is 100% load on all the old nodes. They are obviously doing something, and we need to figure out what. One guess is that these nodes are doing **streaming**, and doing this too quickly - either not isolated via scheduling groups from other work, or isolated but given too many shares. But there can be other things going on that is using up all the CPU - and we need to figure out what this CPU-hog is, rather than guess.

In the monitoring snapshot link you sent, most of the graphs are "null". How do we use them? How can we figure out if during the offending time period (9:00 to 10:00), streaming was active or not, and perhaps at what speed, or if any other background operation was running?

**Comment by tgrabiec** [comment]
On Mon, Sep 9, 2019 at 11:38 AM nyh <notifications@github.com> wrote:

> This is *not* the same issue that happens on a simple node reboot, which
> comes up with an empty cache and slows everything down. In such a case you
> would see low performance, but also low load on most nodes. What we see
> here is 100% load on all the old nodes. They are obviously doing something,
> and we need to figure out what. One guess is that these nodes are doing
> *streaming*, and doing this too quickly - either not isolated via
> scheduling groups from other work, or isolated but given too many shares.
> But there can be other things going on that is using up all the CPU - and
> we need to figure out what this CPU-hog is, rather than guess.
>
> In the monitoring snapshot link you sent, most of the graphs are "null".
> How do we use them? How can we figure out if during the offending time
> period (9:00 to 10:00), streaming was active or not, and perhaps at what
> speed, or if any other background operation was running?
>

SStable reads are more CPU-hungry than cache reads, that could also explain
increase in CPU utilization.

>

**Comment by nyh** [comment]
@tgrabiec yes, but where do these alleged sstable reads (the graphs are missing so I can't check) come from? Since the existing nodes did not lose their cache, they shouldn't have more sstable reads than usual on normal client reads. We can, indeed, have sstable reads for the streaming of the data to the new node. This definitely **should** cause performance to lower while the streaming is going on, but we don't have evidence that this is what is affecting us in this case.

One thing that puzzles me regarding my streaming explanation (so maybe it is just a red herring) is why this becomes a bottleneck in the readers (the old node) but not the writer (the new node). Theoretically, the new node should be doing more work than each of the old nodes sending it data, so only a limited fraction of the CPU on the old nodes should be used for streaming, which should result in a much of the CPU remaining available for the client workload - we shouldn't (?) see performance drop 30-fold (10-fold drop in throughput with 3-fold increase of CPU usage) as we see in these graphs.

**Comment by fruch** [comment]
@nyh here's the live monitor:
http://34.245.138.101:3000/d/detail-3-1/scylla-per-server-metrics?orgId=1&from=1567997321008&to=1568025901640&var-by=instance&var-cluster=&var-dc=All&var-node=All&var-shard=All&var-func=sum

**Comment by nyh** [comment]
@gleb-cloudius I looked at the task queue metrics, `sum(scylla_scheduler_queue_length) by (group)` (there isn't such a thing on the panels, so I had to do it manually in Promethus). Graph attached.

The interesting thing is that there is indeed a huge amount of Streaming CPU work! Starting at 9:00 (which, confusingly, Prometheus calls 6:00...), there is a scheduler queue length of over 4,000 in the streaming scheduling group, for more than half an hour. 

According to @avikivity, when any queue length is over 1,000, we stop polling, and this slows down everything - the requests, the streaming, and everything. Maybe this explains why we see so much Streaming work in the task queue, but so little streaming IO actually being done.
![a](https://user-images.githubusercontent.com/584227/64537301-3d767400-d323-11e9-87c0-7afb4a666359.png)

@asias is it possible that streaming during core replace somehow has too high concurrency on the sending nodes, causing 4,000 continuations to become ready at any given moment? If this is the case then it seems our scheduler can't handle that.

**Comment by tgrabiec** [comment]
There's a related seastar issue: https://github.com/scylladb/seastar/issues/301

**Comment by nyh** [comment]
Thanks @tgrabiec - this issue indeed explains how I/O-bound performance (which is true in our write-only load<1 workload) goes down when there are >1,000 ready continuations. But the question now is why does streaming constantly have more than 4000 ready continuations continuously (for more than half an hour). I'm hoping that @asias can shed some light on the concurrency of "replace node" streaming (on the sending-node side).

**Comment by nyh** [comment]
@asias @tgrabiec I just looked again at the **tltle** of this issue - 1000 keyspaces - and something dawned on me - is it possible we're streaming each keyspace (or each table?) separately, in parallel? This can explain why we have thousands of continuations, and why we didn't see this problem in the past, in workloads with a smaller number of keyspaces.

**Comment by fruch** [comment]
@nyh sorry for mentioning it **only** in the title... I've update it also in the summery

**Comment by fruch** [comment]
@tgrabiec  
so using --max-task-backlog with bigger number then 1000, i.e. bigger then number of keyspaces going to be used ?, is that something used already in one of our productions setup ?

**Comment by tgrabiec** [comment]
Not sure how many tasks per keyspace are actually started. There could be
some range parallelism within the keyspace, in which case you need a
multiple of that. Running the test with say --max-task-backlog=1000000
should validate the theory.

I'm not aware if --max-task-backlog is used in any of our production setups.

**Comment by nyh** [comment]
@fruch it will be a great idea to use @tgrabiec's idea - set max-task-backlog option to very hight, to validate my theory.

If it is validated, I think the fix is to figure out why the concurrency is proportional to the number of keyspaces, and whether we can limit it (e.g., with a semaphore) to some fixed number, e.g, 100 or even 1...

**Comment by fruch** [comment]
@nyh 
here's a live monitor showing twice: 
http://34.245.58.168:3000/d/JMYqeTcWz/scylla-per-server-metrics-nemesis-3-1?orgId=1&from=1568325555760&to=1568394525599&var-by=instance&var-cluster=&var-dc=All&var-node=All&var-shard=All

Do we still want to try out with `--max-task-backlog=1000000` or better to think a different angle for this one ?

**Comment by nyh** [comment]
It turns out that although we do have large **sum** of concurrency, they are not huge on any shard. The sum on all shards on all nodes reaches 4000, but the maximum on any one shard never surpasses 350:

![a1](https://user-images.githubusercontent.com/584227/64919044-ece19980-d7ae-11e9-8cd4-bd8b3f4348c4.png)

So I think we still have no idea what the problem is, and changing `--max-task-backlog` probably won't help :-(

**Comment by asias** [comment]
> It turns out that although we do have large **sum** of concurrency, they are not huge on any shard. The sum on all shards on all nodes reaches 4000, but the maximum on any one shard never surpasses 350:

@nyh Are you saying according to the metrics we actually did not see massive tasks scheduled which you suspected before that streaming generated too many tasks in parallel?

**Comment by slivne** [comment]
we also have this https://github.com/scylladb/scylla/issues/4712

**Comment by fruch** [comment]
seem like we still have similar issue in current run of 3.2:
![image](https://user-images.githubusercontent.com/340979/71129499-41564400-21f8-11ea-8b02-d7e6aa67755f.png)

http://54.246.193.151:3000/d/Z-LW3dfZz/scylla-per-server-metrics-nemesis-3-2?orgId=1&from=1576659760434&to=1576665630020&var-by=instance&var-cluster=&var-dc=All&var-node=All&var-shard=All

this is during decommission, the drop is ~50% which is better what we've seen in 3.1, and all node CPUs is maxed out as before. 

test is still running, we didn't run `terminate_and_replace` yet in this run.

**Comment by dorlaor** [comment]
Worth to realize the root cause - what consumes the cpu? Is it due to
a function of the amount of keyspaces?
Does it happen every time you'll decommission a node in such a setup?

**Comment by fruch** [comment]
Did had a chance yet to drill down further,

But our working assumption is this is connected to the number of keyspaces. since we didn't seen this kind of behavior in other test cases.
And yes, seems like it happened on both decommissions that took place in this run.

**Comment by piodul** [comment]
While reproducing and testing the issue #4712, I observed a similar situation to the one described in this issue. I looked at the metrics related to hints, and observed that new hints start being generated at some point. Roughly at the same time, write requests start to time out.

If I stop cassandra-stress and wait a short while, then all hint traffic stops and those metrics stop changing. After starting cassandra-stress again, the old performance comes back.

I have a theory that the hints are the culprit here, and they start to overload regular cassandra-stress writes. This may cause nodes to appear temporarily not alive, which results in more hints being accumulated.

Below I attach a screenshot of detailed metrics:

![2020-01-20-104733_2560x1440_scrot](https://user-images.githubusercontent.com/4950936/72721819-2f601c00-3b7d-11ea-9d66-defdf0194794.png)
![2020-01-20-120431_2560x1440_scrot](https://user-images.githubusercontent.com/4950936/72721829-3555fd00-3b7d-11ea-91e7-34346bc99288.png)

Some hint-related graphs:

![2020-01-20-113319_2560x1440_scrot](https://user-images.githubusercontent.com/4950936/72721864-4bfc5400-3b7d-11ea-8a20-ea0825b78fa5.png)

The test was performed on Scylla version 3.1.0.rc2.

**Comment by slivne** [comment]
@piodul / @fruch I need to understand if this is related to hintted-handoff or not
@fruch we need to run the same scenario with hinted handoff disabled and then to check if the drop still exists - if it doesn't hintted-handoff is to blame and we can mark this issue accordingly and allow piodul to tackle it.

If this happens without HH then please provide the info and we will allow someone else to tackle this - while piodul will split this ticket apart for HH issue.

**Comment by fruch** [comment]
@slivne 
running now:
https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/longevity/job/longevity-multi-keyspaces-60h/3/

**Comment by asias** [comment]
In "Performance dropped in twice after node decomission #5495 ", we saw big reactor stalls. However, we do not see big stall in the replacing operations. Since with replacing operations, no token ring changes, there is no calculation of pending ranges update. 

```
 [ 1000_ks_terminate_and_replace_throughput_drop]$ grep -ri 'Reactor stalled'  .
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T03:39:12+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 651 ms on shard 18.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T03:43:11+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 652 ms on shard 13.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T03:44:23+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 652 ms on shard 9.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:05:51+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 651 ms on shard 22.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:07:42+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 650 ms on shard 19.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:10:04+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 651 ms on shard 5.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:16:10+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 650 ms on shard 28.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:27:00+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 653 ms on shard 16.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:27:16+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 651 ms on shard 6.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:27:17+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 651 ms on shard 26.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:27:49+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 653 ms on shard 21.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:33:49+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 651 ms on shard 10.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-2/database.log:2019-09-09T03:33:33+00:00  ip-10-0-8-59 !INFO    | scylla: Reactor stalled for 652 ms on shard 23.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-2/database.log:2019-09-09T03:42:07+00:00  ip-10-0-8-59 !INFO    | scylla: Reactor stalled for 652 ms on shard 4.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-2/database.log:2019-09-09T04:40:17+00:00  ip-10-0-8-59 !INFO    | scylla: Reactor stalled for 653 ms on shard 24.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-2/database.log:2019-09-09T04:42:48+00:00  ip-10-0-8-59 !INFO    | scylla: Reactor stalled for 650 ms on shard 21.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-2/database.log:2019-09-09T06:01:53+00:00  ip-10-0-8-59 !INFO    | scylla: Reactor stalled for 652 ms on shard 28.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-2/database.log:2019-09-09T06:01:58+00:00  ip-10-0-8-59 !INFO    | scylla: Reactor stalled for 652 ms on shard 2.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-6/database.log:2019-09-09T03:42:54+00:00  ip-10-0-56-177 !INFO    | scylla: Reactor stalled for 653 ms on shard 29.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-6/database.log:2019-09-09T03:43:00+00:00  ip-10-0-56-177 !INFO    | scylla: Reactor stalled for 651 ms on shard 28.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-6/database.log:2019-09-09T03:59:13+00:00  ip-10-0-56-177 !INFO    | scylla: Reactor stalled for 651 ms on shard 13.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-6/database.log:2019-09-09T04:01:09+00:00  ip-10-0-56-177 !INFO    | scylla: Reactor stalled for 651 ms on shard 16.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-6/database.log:2019-09-09T04:16:41+00:00  ip-10-0-56-177 !INFO    | scylla: Reactor stalled for 651 ms on shard 28.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-6/database.log:2019-09-09T04:17:31+00:00  ip-10-0-56-177 !INFO    | scylla: Reactor stalled for 651 ms on shard 5.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-6/database.log:2019-09-09T04:18:53+00:00  ip-10-0-56-177 !INFO    | scylla: Reactor stalled for 651 ms on shard 9.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-7/database.log:2019-09-09T04:17:53+00:00  ip-10-0-107-93 !INFO    | scylla: Reactor stalled for 654 ms on shard 0.
```

**Comment by fruch** [comment]
I'm running with the formal 3.2.0, with hinted handoff disabled, doing only TerminateAndReplaceNode

http://34.241.207.103:3000/d/q4I08GsWk/scylla-per-server-metrics-nemesis-3-2?orgId=1&from=1579556598406&to=1579588910028&var-by=instance&var-cluster=&var-dc=All&var-node=All&var-shard=All&var-sct_tags=DisruptionEvent
 
![image](https://user-images.githubusercontent.com/340979/72779851-eca64e80-3c25-11ea-8551-f3915d4c631a.png)

seems a bit more saner then the 50%-90% percent drop in throughput we seen.
@slivne @piodul, the cluster if still up and running. if you want more information out of it to confirm.

**Comment by piodul** [comment]
@fruch I noticed that during the first replace there is a sharp drop in number of requests served. Could this be caused by prepare_write_cmd ending there, and the nemesis starting immediately? I can see that the prepare_stress_cmd uses more threads than the stress_cmd, so that would explain why the throughput is higher before the drop.

Indeed, the performance goes back up to the original level after a second replacement, but we don't know what would be the performance before the first replacement. We risk here that the performance drop after replacement is idempotent, i.e. when replacing for the second time, performance drop stays the same. A better test would be to let the regular stress command run for a short while, then run the replacement nemesis. Could it be easily done via SCT?

By the way, this graph shows us that the original performance drop _during_ replacement does not occur when hints are disabled, which is what issue #4712 was about. The comb-like parts of the graph are probably caused by nodetool repair.

**Comment by asias** [comment]
In "Performance dropped in twice after node decomission #5495 ", we saw big reactor stalls. However, we do not see big stall in the replacing operations. Since with replacing operations, no token ring changes, there is no calculation of pending ranges update. 

```
 [ 1000_ks_terminate_and_replace_throughput_drop]$ grep -ri 'Reactor stalled'  .
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T03:39:12+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 651 ms on shard 18.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T03:43:11+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 652 ms on shard 13.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T03:44:23+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 652 ms on shard 9.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:05:51+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 651 ms on shard 22.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:07:42+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 650 ms on shard 19.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:10:04+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 651 ms on shard 5.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:16:10+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 650 ms on shard 28.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:27:00+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 653 ms on shard 16.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:27:16+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 651 ms on shard 6.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:27:17+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 651 ms on shard 26.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:27:49+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 653 ms on shard 21.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-1/database.log:2019-09-09T04:33:49+00:00  ip-10-0-209-157 !INFO    | scylla: Reactor stalled for 651 ms on shard 10.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-2/database.log:2019-09-09T03:33:33+00:00  ip-10-0-8-59 !INFO    | scylla: Reactor stalled for 652 ms on shard 23.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-2/database.log:2019-09-09T03:42:07+00:00  ip-10-0-8-59 !INFO    | scylla: Reactor stalled for 652 ms on shard 4.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-2/database.log:2019-09-09T04:40:17+00:00  ip-10-0-8-59 !INFO    | scylla: Reactor stalled for 653 ms on shard 24.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-2/database.log:2019-09-09T04:42:48+00:00  ip-10-0-8-59 !INFO    | scylla: Reactor stalled for 650 ms on shard 21.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-2/database.log:2019-09-09T06:01:53+00:00  ip-10-0-8-59 !INFO    | scylla: Reactor stalled for 652 ms on shard 28.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-2/database.log:2019-09-09T06:01:58+00:00  ip-10-0-8-59 !INFO    | scylla: Reactor stalled for 652 ms on shard 2.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-6/database.log:2019-09-09T03:42:54+00:00  ip-10-0-56-177 !INFO    | scylla: Reactor stalled for 653 ms on shard 29.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-6/database.log:2019-09-09T03:43:00+00:00  ip-10-0-56-177 !INFO    | scylla: Reactor stalled for 651 ms on shard 28.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-6/database.log:2019-09-09T03:59:13+00:00  ip-10-0-56-177 !INFO    | scylla: Reactor stalled for 651 ms on shard 13.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-6/database.log:2019-09-09T04:01:09+00:00  ip-10-0-56-177 !INFO    | scylla: Reactor stalled for 651 ms on shard 16.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-6/database.log:2019-09-09T04:16:41+00:00  ip-10-0-56-177 !INFO    | scylla: Reactor stalled for 651 ms on shard 28.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-6/database.log:2019-09-09T04:17:31+00:00  ip-10-0-56-177 !INFO    | scylla: Reactor stalled for 651 ms on shard 5.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-6/database.log:2019-09-09T04:18:53+00:00  ip-10-0-56-177 !INFO    | scylla: Reactor stalled for 651 ms on shard 9.
./longevity-1000-keyspaces-3-1-db-node-e15330fe-7/database.log:2019-09-09T04:17:53+00:00  ip-10-0-107-93 !INFO    | scylla: Reactor stalled for 654 ms on shard 0.
```

> I'm running with the formal 3.2.0, with hinted handoff disabled, doing only TerminateAndReplaceNode
> 
> http://34.241.207.103:3000/d/q4I08GsWk/scylla-per-server-metrics-nemesis-3-2?orgId=1&from=1579556598406&to=1579588910028&var-by=instance&var-cluster=&var-dc=All&var-node=All&var-shard=All&var-sct_tags=DisruptionEvent
> 
> ![image](https://user-images.githubusercontent.com/340979/72779851-eca64e80-3c25-11ea-8551-f3915d4c631a.png)
> 
> seems a bit more saner then the 50%-90% percent drop in throughput we seen.
> 
> Logs:
> 
> ```
> +--------------------------------------------------------------------------------------------------------------------------------------------+
> |  Collected logs by test-id: 8972866b-bd94-4ca6-a173-ca66379823c4 Directory: /home/fruch/sct-results/20200121-100847-130691/collected_logs  |
> +-------------+------------------------------------------------------------------------------------------------------------------------------+
> | Cluster set | Link                                                                                                                         |
> +-------------+------------------------------------------------------------------------------------------------------------------------------+
> | db-cluster  | https://cloudius-jenkins-test.s3.amazonaws.com/8972866b-bd94-4ca6-a173-ca66379823c4/20200121_102756/db-cluster-8972866b.zip  |
> | monitor-set | https://cloudius-jenkins-test.s3.amazonaws.com/8972866b-bd94-4ca6-a173-ca66379823c4/20200121_102756/monitor-set-8972866b.zip |
> | loader-set  | https://cloudius-jenkins-test.s3.amazonaws.com/8972866b-bd94-4ca6-a173-ca66379823c4/20200121_102756/loader-set-8972866b.zip  |
> | sct-runner  | https://cloudius-jenkins-test.s3.amazonaws.com/8972866b-bd94-4ca6-a173-ca66379823c4/20200121_102756/sct-runner-8972866b.zip  |
> +-------------+------------------------------------------------------------------------------------------------------------------------------+
> 
> to get monitor data: 
> hydra investigate show-monitor 8972866b-bd94-4ca6-a173-ca66379823c4
> ```


Thanks! 

There are two drops (24:00 to 01:00  and 03:30 to 05:30). Did the test run two replace operations?

The CPU hit 100% during the replace operation as the streaming kicks in. The performance drop is expected. Also, the drop was not constant during the period. It dropped up to 30K and went up to 60K, fluctuating during the whole replace operation.

**Comment by slivne** [comment]
@piodul / @fruch / @asias 
- 90% drop is due to hinted handoff 
- with hinted handoff disabled there is a non constant drop to 50% of the throughput - but that is going up and down.

- If this is true - lets mark this issue as hintted handoff related (as its the cause for 90% drop) and create a different issue of what happens when hintted handoff is disabled and we replace a node and we will look at the two in parallel.

**Comment by piodul** [comment]
@asias, @fruch I opened a new issue about the fluctuating performance in @fruch's most recent test run. Feel free to update any details if something is missing/wrong: #5647

**Comment by slivne** [comment]
@piodul whats the current status - after moving hints to its own verb and making it on the correct scheduling group is this any better ?

**Comment by piodul** [comment]
@slivne I think it have helped. While I did encounter a performance drop when running a (different) longevity test with fixed Scylla (more details in #4712), it happened only during repair of a keyspace that had materialized views in it. During repair of other keyspaces, the performance was reasonable. This longevity test does not use MVs, so I would now expect performance to be better here.

**Comment by slivne** [comment]
@fruch / @roydahan - without MV's can you please rerun the test and provide the info if this is solved - if so great and lets close this. 

If there is still a throughput drop that is more than 10% but less than 20% please close this issue (after all 90% drop is fixed :) ) and open a new one and we will check that out as well.

**Comment by slivne** [comment]
looking at https://github.com/scylladb/scylla/issues/4712#issuecomment-626365652 I think we may still need to fix something

**Comment by slivne** [comment]
@roydahan please provide updated info from 4.4 run so we can know if we can close this
